DOC-161 Added plugins directory, pages, and added to sidebar for 24.2 (#4170)
EventStore/EventStore: docs/plugins/User-certificates-plugin.md

# User Certificates Plugin 
The User Certificates plugin enhances security by allowing authentication through both a username/password and an x509 user certificate. User certificates work across any cluster with a shared root Certificate Authority (CA) with the user's certificate. This means that you can have a single user certificate that is valid across multiple clusters.

::: tip
The User Certificates plugin is included in commercial versions of EventStoreDB.
:::

> # Setup Instructions

Reviewer: ```suggestion
## Setup Instructions
```

1. Download the User Certificates plugin.

> 2. Place it in the `.\plugins` directory within your EventStoreDB installation. 

Reviewer: ```suggestion
2. Place it in the `.\plugins` directory within your EventStoreDB installation directory. 
```

> 3. In the EventStoreDB `.\config` directory, add the following configuration to `.\config\plugin-config.json`:

Reviewer: ```suggestion
3. Find the file `.\config\plugin-config.json` in the EventStoreDB `.\config` directory, add the following configuration:
```

```
{
	"EventStore": {
		"Plugins": {
			"UserCertificates": {
				"Enabled": true
			}
		}
	}
}
```

After installation and configuration, the server logs an entry confirming the plugin's activation:

```
[13260, 8,13:59:10.445,INF] UserCertificatesPlugin: user X.509 certificate authentication is enabled
```

## Connecting with a User Certificate

Use the following command as an example to connect using `curl`: 

```
curl -i https://localhost:2113/streams/%24all --cert user-admin.crt --key user-admin.key
```

## Creating a User Certificate

The user certificate must adhere to the following requirements:

- The certificate has a root CA in common with the node certificate.
- The root CA that they have in common is trusted by the node.
- The CN of the certificate is the same as the username.
- The certificate has the ClientAuth EKU, and not the ServerAuth EKU.
- The certificate must be in date.
- The CN is the username of a user that exists in the database.

> > Note: Certificates can also authenticate `admin` and `ops` users.

Reviewer: this `>` will just make a blockquote. I suggest turning this into a `::: tip` box

Author: Good eye @rsowald

    00d0  42 3c 11 f1 f1 2c 93 9d  6d 8b b8 ec eb c6 e5 e7
    00e0  a4 9a d4 59 fb ff b7 ba  e1 33 c2 f8 09 52 80 c6
    00f0  95 95 51 35 08 1a 65 41  76 bc c2 62 2b 71 b3 65
Non-root Certificate
Key Id Hash(rfc-sha1): 8a3c17e6662711175142740cfeb0f52d2bdd3596
Key Id Hash(sha1): 87a30feaaeabd616f4dad92a9f1b44a43e2d0b52
Subject Key Id (precomputed): 01fc7f1e364f1767b759dae23d43bbb1cbf2d75deabe605737969489b8de201d
Key Id Hash(bcrypt-sha1): b808b4e710e4f6f158a837155838a295cfbb17bf
Key Id Hash(bcrypt-sha256): eaf799bb5f3d57dbb3afc0f5357a8723574c8c254bf159fce1c95d76f6085ae5
Key Id Hash(md5): fbb024d546e2039af7f3e01c71278313
Key Id Hash(sha256): ca10f7cb5af9bf9dff8f2ad43c3cbdb0647da1173631018cb8551812c425d575
Key Id Hash(pin-sha256): MZu7Waanp4kbJ/Hx3/tpgMb6L3GnXB3PrMY+9MDBmSg=
Key Id Hash(pin-sha256-hex): 319bbb59a6a7a7891b27f1f1dffb6980c6fa2f71a75c1dcfacc63ef4c0c19928
Cert Hash(md5): 1719850bd6f5f09c3610e4c33d13f7c2
Cert Hash(sha1): 12c5293973bda1df2c74ab280d492135171c8ab6
Cert Hash(sha256): a37767096f10697fb0fcf94a3872d0984aeed4754a295c5320679876e00def32
Signature Hash: 6d922badaba2372070f13c69b620286262eab1d8d2d2156a271a1d73aaaf64e4
```

</details>

## Restrictions

> There is currently a limitation with this plugin that requests authenticated with user certificates cannot be forwarded between nodes.

Reviewer: ```suggestion
There is currently a limitation with this plugin: requests authenticated with user certificates cannot be forwarded between nodes.
```

This affects Writes, Persistent Subscription operations, and Projections operations. These requests will need to be performed on the Leader node only.

The User Certificates plugin cannot be used alongside any of the other authentication plugins, such as the LDAPS plugin or the OAuth plugin.

## Troubleshooting

### Plugin Not Loaded
Ensure the plugin is correctly placed in the subdirectory within the `plugins` directory of your server's installation folder. If the `plugins` directory or the specific plugin subdirectory doesn't exist, you need to create them and place the plugin binaries there. To verify the plugin setup:

1. Go to the EventStoreDB installation directory where the executable is located. 
2. Ensure there's a directory called `plugins`. If not, create one. 

> 3. Inside `plugins`, create a subfolder for the OTLP Exporter Plugin (e.g.,`EventStore.Auth.UserCertificates` or any name you prefer), if it's missing.

Reviewer: ```suggestion
3. Inside `plugins`, create a subfolder for the User Certificates Plugin (e.g.,`EventStore.Auth.UserCertificates` or any name you prefer), if it's missing.
```

4. Place the plugin binaries in this subfolder.

You can verify which plugins have been found and loaded by looking for the following logs:

```
[12824, 1,17:30:04.921,INF] Plugins path: "C:\eventstoredb-24.2.0-rc.1-ce-windows.x64\plugins"
[12824, 1,17:30:04.922,INF] Adding: "C:\eventstoredb-24.2.0-rc.1-ce-windows.x64\plugins" to the plugin catalog.
[12824, 1,17:30:04.924,INF] Adding: "C:\eventstoredb-24.2.0-rc.1-ce-windows.x64\plugins\EventStore.Auth.UserCertificates" to the plugin catalog.
[12824, 1,17:30:05.253,INF] Loaded SubsystemsPlugin plugin: "user-certificates" "24.2.0.1".
```

### Plugin Not Started

> The plugin has to be configured to start to authenticate user requests.

Reviewer: ```suggestion
The plugin has to be configured to start authenticating user requests.
```
